2019 minikube roadmap! (#3488)
kubernetes/minikube: docs/contributors/roadmap.md

Reviewer: Was this goal removed ? It would be good if libmachine could be made more engine-agnostic.

Author: Yes, at least for 2019. I'm not convinced that there is much merit in adding another layer of abstraction here, unless it's a migration path to something better.

In 2018, we changed our upstream project to machine-drivers, though I'm not yet convinced of that projects long-term viability unless activity picks up.

Reviewer: Machine drivers organisation was mostly created to host _drivers_ blocked by upstream, such as the KVM driver (forked and used by minikube). Its possibility to develop API seems questionable...

At this point in time, there is **no** difference to `libmachine` itself. So the change is mostly superficial, the bigger issue what is going to happen with the two libvirt machine drivers used by minikube ?

Reviewer: Was this goal removed ? There's a lot of confusion about the KVM drivers, and "upstream" is gone.

Author: Thanks for bringing this up. I've added it to the technical debt section.

# minikube roadmap (2019)

This roadmap is a living document outlining the major technical improvements which we would like to see in minikube during 2019. Please send a PR to suggest any improvements to it.

## (#1) Make minikube the easiest way for developers to learn Kubernetes

- Single step installation

>   - Users should not have to seperately install supporting binaries

Reviewer 2: ```suggestion
  - Users should not have to separately install supporting binaries
```

Author: Done.

- Creation of a user-centric minikube website for installation & documentation

## (#2) Diversify the minikube community

- Add documentation for new minikube contributors

> - Ensure that all decisions are ratified publically by the minikube community

Reviewer 2: ```suggestion
- Grow the number of maintainers and increase the involvement of the minikube community in planning and decision making 
```

Author: Done.

## (#3) Make minikube robust and debuggable

- Add pre-flight error checks for common connectivity issues 
- Add pre-flight error checks for common configuration errors
- Mark features & options not covered by continuous integration as `experimental`

> - Improve the `status` command so that it can diagnose common environmental issues

Reviewer 2: ```suggestion
- Improve `minikube status` so that it can diagnose common environmental issues
```

Author: Done.

- Make minikube usable offline

## (#4) Official multi-node/multi-cluster support

> - Rebrand profiles as multi-cluster support

Reviewer 2: I think those are two separate use cases. 
- multiple single node clusters for different projects for example - 1 profile each 
- 1 multinode cluster - profiles might be useful for this but "rebranding" doesn't make sense for me

Author: Correct, they are separate use cases. Both of which we should be able to address.

Other parts of the Kubernetes community refers to what we have in profiles as multi-cluster, so I figure we should use common terminology.

> - Integrate KIND for multi-node support

Reviewer 2: I'd pull this out, it's an implementation detail and I'm not convinced that KIND will be the solution here

Reviewer: Sounds easier to build on the current `kubeadm`, and make it easier to add multiple nodes beyond the first ?

Author: Done.

- Add commands to add/remove nodes within an existing cluster

## (#5) Improve minikube performance

> - Add support for lighter-weight deployment methods, such as container-based (LXD, Docker) or chroot

Reviewer: This seems to contradict previous decisions, like dropping boot2docker and localkube for "standard" solutions ?

It will be hard for `minikube` to compete with vendor solutions, such as microk8s (for LXD) or Docker Desktop.

Author: I've adjusted this a bit. I don't believe for instance, that chroot or LXD support needs to contradict previous decisions. 

minikube already competes/cooperates with other vendor solutions. This is more to address the use cases where folks want to use minikube, but can't because it conflicts with other requirements. For instance, many corporate environments (like my own!) forbid VM's, and running VM's on Chromebooks is awkward at best. These users should still be able to use minikube, at least on Linux.

People already use "none" to work around the fact that minikube has nothing better to offer them, and it scares the hell out me. We can do better.

# minikube roadmap (2019)

This roadmap is a living document outlining the major technical improvements which we would like to see in minikube during 2019, divided by how they apply to the minikube [(guiding principles)[principles.md]

Please send a PR to suggest any improvements to it.

> ## (#1). User-friendly and accessible

Reviewer 2: ```suggestion
## (#1) User-friendly and accessible
```

- Increase community involvement in planning and decision making
- Make the continuous integration and release infrastructure publicly available
- Double the number of active maintainers

## (#3) Cross-platform

- Simplified installation process across all supported platforms
- Users should never need to separately install supporting binaries

## (#4) Support all Kubernetes features

- Add multi-node support

## (#5) High-fidelity

- Reduce guest VM overhead by 50%
- Disable swap in the guest VM

## (#6) Compatible with all supported Kubernetes releases

- Continuous Integration testing across all supported Kubernetes releases
- Automatic PR generation for updating the default Kubernetes release minikube uses

## (#7) Support for all Kubernetes-friendly container runtimes

- Run all integration tests across all supported container runtimes
- Support for Kata Containers (help wanted!)

## (#8) Stable and easy to debug

- Pre-flight error checks for common connectivity and configuration errors
- Improve the `minikube status` command so that it can diagnose common issues
- Mark all features not covered by continuous integration as `experimental`

> - Replace built-in machine drivers (virtualbox, kvm2) with their upstream equivalents

Reviewer: AFAIK, the "virtualbox" driver is already built-in (to machine) and "kvm2" doesn't have an upstream (from docker).
Replacing kvm2 with kvm sounds like a step backwards, better to continue to evolve the minikube KVM driver ?

Author: Thanks. This was supposed to say hyperkit and kvm2. I'm just going to remove this from the technical roadmap, since it's strictly about technical debt reduction and is not terribly inspirational or important to end users.

The idea is to push our driver improvements upstream so that we may share improvements with minishift and other consumers:

kvm -> https://github.com/machine-drivers/docker-machine-driver-libvirt
hyperkit -> https://github.com/machine-drivers/docker-machine-driver-hyperkit

But only if doing so moves minikube forwards rather than backwards.

Thanks for the input!

> - Remove dependency on boot2docker (deprecated)

Reviewer: I think the dependency on boot2docker.iso has _already_ been removed, and replaced with custom minikube.iso ?
So this goal doesn't make sense, unless also wanting to remove libmachine (since it has also been deprecated...)

Author: Removed.
